### PR TITLE
Prevent rules incorrectly returning conflicting fixes to same position

### LINKED ISSUE #2827: Snowflake - Fix Error when semicolon is preceeded by space
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

SQLFluff identifies fixable error - and fails to fix it

### Expected Behaviour

Semicolon snaps to end of statement

### Observed Behaviour

No semicolon identified at end of statement, unable to fix

### How to reproduce

sqlfluff fix test.sql --dialect=snowflake

```
SET foo = (SELECT foo FROM foo.foo) ;

```

### Dialect

Snowflake

### Version

Current HEAD of master

### Configuration

Default config - but snowflake

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by barrywhart**
### Brief summary of the change made
Fixes #2827 

Changes:
* Fixes L052 (as noted in the issue)
* Fixes L036, L050,L053 (which had other issues uncovered by CI or internal linter checks)
* Changes core linter behavior for rules returning multiple fixes with same anchor segment.
  * Old behavior: Arbitrarily applies one of the fixes, silently discards other fixes with the same anchor. Note this means fixes are not atomic -- bad!!
  * New behavior: Logs a warning and ignores the whole set of fixes. (During automated tests, raises an error rather than logging.)
 

### Pull Request checklist
- [X] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

File: src/sqlfluff/core/linter/linter.py
Comment (barrywhart) [inline src/sqlfluff/core/linter/linter.py:526]: New error detection code.

File: src/sqlfluff/core/parser/segments/base.py
Comment (barrywhart) [inline src/sqlfluff/core/parser/segments/base.py:1056]: Delete this comment now that we're dealing with them (by warning and discarding)

File: src/sqlfluff/rules/L052.py
Comment (barrywhart) [inline src/sqlfluff/rules/L052.py:180]: The bug was here.

File: src/sqlfluff/rules/L052.py
Comment (barrywhart) [inline src/sqlfluff/rules/L052.py:229]: This code was practically identical to the buggy code. I refactored it to use the same helper function, thus it gets the same fix.

**Comment by WittierDinosaur** [comment]
New error seems to have blown up L036? Maybe more underlying bugs?

**Comment by barrywhart** [comment]
Yes, L036 has a bug as well. Interesting that we hadn't noticed it before. I'll try and fix that in the same PR, as long as it's not a big nasty fix.

**Comment by barrywhart** [comment]
Looking at one of the L036 failures, it is indeed returning multiple fixes with the same anchor. In this case, they are identical (both are `delete`s). That's why we haven't noticed previously. (If the fixes were different, we likely would've noticed, as with L052.) Should be a simple fix. 🤞

**Comment by barrywhart** [comment]
Ok, I fixed the L036 bug. It appears L001 has a "duplicate anchors" bug as well!

**Comment by tunetheweb** [comment]
> Ok, I fixed the L036 bug. It appears L001 has a "duplicate anchors" bug as well!

06C57055-C735-4B09-AA2F-6CDF6E392A37

**Comment by barrywhart** [comment]
It turns out L001 was okay. In one test, it was returning two deletions with segments that had the same position info, but were different objects. I updated the PR to use a new class, `IdentitySet`, that checks for membership by object identity rather than equality.

**Comment by barrywhart** [comment]
@tunetheweb, @WittierDinosaur: Ok, ready for review!!

File: src/sqlfluff/core/linter/linter.py
Comment (barrywhart) [inline src/sqlfluff/core/linter/linter.py:527]:
```suggestion
                            f"the same anchor. This is not supported, so the "
```

File: src/sqlfluff/rules/L052.py
Comment (barrywhart) [inline src/sqlfluff/rules/L052.py:242]: New function extracted from existing code. This is where we fixed the L052 bug.

File: src/sqlfluff/rules/L052.py
Comment (barrywhart) [inline src/sqlfluff/rules/L052.py:256]: This is the heart of the bug fix.

File: test/conftest.py
Comment (barrywhart) [inline test/conftest.py:252]: Extends the recently added fixture to make tests fail on fixes with duplicate anchors.

File: src/sqlfluff/core/linter/linter.py
Comment (tunetheweb) [inline src/sqlfluff/core/linter/linter.py:334]: Seems a bit of a pointless function? One line of code to log the param, to save you typing one line of code to log the param? The similar `_warn_unfixable` at least has the error message in the function with only the variable part passed but here the whole message is passed.

File: src/sqlfluff/core/linter/linter.py
Comment (tunetheweb) [inline src/sqlfluff/core/linter/linter.py:530]: I presume this has to be unique per rule?

i.e. if you have this:

```sql
SELECT DATE_ADD(date,col1)
```

And you had to change that to:

```sql
SELECT DATE_ADD(date, "col1")
```

That is add a space AND change `col1` to be quoted, would that be same anchor point? But as it's different rules there shouldn't be a problem? Or do all rules fixes get queued up to one big set?

I know you say two fixes for same anchor points don't work, but just wondering if they do in some cases and we're now going to break that?

Should this throw a CRITICAL exception so my rules checker will see that?

File: test/fixtures/rules/std_rule_cases/L052.yml
Comment (tunetheweb) [inline test/fixtures/rules/std_rule_cases/L052.yml:475]: The simpler statement `SELECT 1  ;` worked before and autofixed.

So that clearly means it went down a different code path (cause it's actually `SELECT<indent>1<dedent> ;` and those `indents/dendents` seemed to be enough to allow it to auto fix with the old code.

Therefore, as it's going down a different code path (or at least used to!), I think we should add that as a test case too.

```suggestion
test_fail_whitespace_after_simple_select:
  fail_str: |
    SELECT 1 ;
  fix_str: |
    SELECT 1;
 
test_fail_whitespace_after_snowflake_set:
  fail_str: |
    SET foo = (SELECT foo FROM foo.foo) ;
  fix_str: |
    SET foo = (SELECT foo FROM foo.foo);
  configs:
    core:
      dialect: snowflake
```

**Review by tunetheweb — CHANGES_REQUESTED** [review]
You're the expert here, but can't see any flags.

Do have a couple of comments/questions and one suggested additional test case.

File: src/sqlfluff/core/linter/linter.py
Comment (barrywhart) [reply src/sqlfluff/core/linter/linter.py:334]: This function exists primarily in order to let us monkeypatch it at runtime (replacing it with a function that raises an exception).

File: src/sqlfluff/core/linter/linter.py
Comment (barrywhart) [reply src/sqlfluff/core/linter/linter.py:530]: Unique per rule, yes. The linter runs one rule at a time, applies all the fixes, then runs the next rule, and this check occurs within each iteration of the rule loop.

I could change the log level if you like. This is similar to the recent "parse after fix", which currently log warnings. If we change this, should we consider changing that as well?

**Comment by barrywhart** [comment]
@tunetheweb: I added the test case you suggested and replied to your other comments. Let me know what you think.

File: src/sqlfluff/core/linter/linter.py
Comment (tunetheweb) [reply src/sqlfluff/core/linter/linter.py:334]: Should we add a comment to that effect?

File: src/sqlfluff/core/linter/linter.py
Comment (tunetheweb) [reply src/sqlfluff/core/linter/linter.py:530]: I think I need to update my script to look for more than just CRITICAL log lines if we want to catch these things. Otherwise they’re silently hidden and we only get notice when users raise them.

Or maybe we should consider them critical?

We could also use that key “CRITICAL” word in the log, so my script catches, but then a bit weird that it’s just a warning log level if we do that. It mightn’t be a bad idea to do that before merging, let one run of CI go through, then revert it back assuming no issues. Longer term I should just fix my script though to look for more key phrases we want it to catch.

Comment (barrywhart) [reply src/sqlfluff/core/linter/linter.py:530]: I'm fine with elevating it to CRITICAL. Since these are bugs in the rules, I don't mind drawing more attention to this.

Comment (tunetheweb) [reply src/sqlfluff/core/linter/linter.py:530]: Cool let’s do that and see if my script picks up anything!

File: src/sqlfluff/core/linter/linter.py
Comment (barrywhart) [reply src/sqlfluff/core/linter/linter.py]: Will do.

File: src/sqlfluff/core/linter/linter.py
Comment (barrywhart) [reply src/sqlfluff/core/linter/linter.py:530]: Changing this now.

**Comment by barrywhart** [comment]
@tunetheweb: Log level increased to `critical` for this message and the "parse after fix" message. Also added comment explaining why the one-line logging function exists.

File: test/fixtures/rules/std_rule_cases/L052.yml
Comment (tunetheweb) [inline test/fixtures/rules/std_rule_cases/L052.yml:473]:
```suggestion

```

**Comment by tunetheweb** [comment]
Oh it found some effors! Good call to make critical.

**Comment by barrywhart** [comment]
@tunetheweb: Whaaa, now more work to do? 😄 I might wait and look at this tomorrow. Stamping out bugs is 😎 , though! Teamwork!

File: src/sqlfluff/core/parser/segments/base.py
Comment (barrywhart) [inline src/sqlfluff/core/parser/segments/base.py:1017]: Below, I reworked the core `apply_fixes()` logic:
* Consumes a dictionary of `AnchorEditInfo` rather than a list of fixes
* Add the ability to handle `create_before` and `create_after` the same anchor (new requirement for L053)

File: src/sqlfluff/core/parser/segments/base.py
Comment (barrywhart) [inline src/sqlfluff/core/parser/segments/base.py:1065]: Note how this tricky "end of loop" logic all goes away now that we're using a dictionary instead.

File: src/sqlfluff/rules/L036.py
Comment (barrywhart) [inline src/sqlfluff/rules/L036.py:298]: @tunetheweb: I'd like your thoughts on a question. This PR adds a new feature that prohibits fixes with the same anchor with one exception: It's okay to have 2 fixes, one `create_before` and one `create_after`.

Should we also allow multiple **deletes** of the same segment? There's no ambiguity there, and it would avoid needing to "fix" this rule as well as L053.

It's kind of a philosophical question. By allowing this exception, we make it easier on rule writers. On the other hand, we're letting them be a bit sloppy. 🤷‍♂️ I could be convinced either way.

File: src/sqlfluff/rules/L050.py
Comment (barrywhart) [inline src/sqlfluff/rules/L050.py:76]: If we allowed multiple deletions, we wouldn't need this change.

File: src/sqlfluff/rules/L039.py
Comment (tunetheweb) [inline src/sqlfluff/rules/L039.py:132]: Why is this code in L039? Feels like core code that should be in BaseRule.

File: src/sqlfluff/rules/L052.py
Comment (tunetheweb) [inline src/sqlfluff/rules/L052.py:252]: This comment feels more like an explanation for this PR, rather than something that needs to be in the code.

```suggestion
```

File: test/fixtures/rules/std_rule_cases/L053.yml
Comment (tunetheweb) [inline test/fixtures/rules/std_rule_cases/L053.yml:81]: Why using `\n` when that's not what's used for the `fail_str`? Shouldn't it be consistent? Would also make the initial space look like "bad".

File: test/fixtures/rules/std_rule_cases/L053.yml
Comment (tunetheweb) [inline test/fixtures/rules/std_rule_cases/L053.yml:94]: Ditto.

**Review by tunetheweb — CHANGES_REQUESTED** [review]
Still missing some coverage too.

File: src/sqlfluff/rules/L036.py
Comment (tunetheweb) [reply src/sqlfluff/rules/L036.py:298]: Hmmm.... I think safer to not allow that. In theory it shouldn't be needed so the developer ease is not a strong enough argument in my mind. And curious why L036 currently does it? Looked at the code but wasn't immediately apparent to me.

File: src/sqlfluff/core/linter/linter.py
Comment (barrywhart) [reply src/sqlfluff/core/linter/linter.py:535]: Will do.

File: src/sqlfluff/core/parser/segments/base.py
Comment (barrywhart) [reply src/sqlfluff/core/parser/segments/base.py:90]: I'll update it to something similar as you suggest.

As currently written, 0 will never occur because we only call `is_valid` if there are fixes. But I'd prefer to treat 0 as valid because it's harmless and gives us a bit of future proofing.

File: src/sqlfluff/rules/L036.py
Comment (barrywhart) [reply src/sqlfluff/rules/L036.py:298]: There's not a "good" reason, and I am not that familiar with L036 details, but basically, there are a couple places that delete unnecessary whitespace, and there was no bookkeeping, so in some cases, the same whitespace gets deleted twice.

Happy to leave the fix checker "as is" for now (i.e. not allow multiple deletes).

File: src/sqlfluff/rules/L039.py
Comment (barrywhart) [reply src/sqlfluff/rules/L039.py:132]: We don't want `BaseRule` to discard when multiple fixes have the same anchor, because we consider it an error (the `is_valid` check).

L039 had duplicate anchors on this new test case (extracted from one of the `.sql` fixtures):
```
test_excess_space_cast:
  fail_str: |
    select
        '1'    ::   INT as id1,
        '2'::int as id2
    from table_a
  fix_str: |
    select
        '1'::INT as id1,
        '2'::int as id2
    from table_a
```

L039 wants to make two fixes to the line `'1'    ::   INT as id1,`.
1. Replace excessively long whitespace with a single whitespace: "    " -> " " (2 places)
2. Entirely **remove** the whitespace around the `::`.

Thus, it's trying to both **replace** and **delete** the same whitespace. If we return both fixes, the core linter will (appropriately) complain and discard **both** fixes. This bookkeeping ensures that both get fixed, but it's split across two passes through the linter loop. There may be a smarter way to do this, but this approach seems reasonable. I'm trying really hard not to do big rewrites of existing rule code during these PRs -- the goal is to eliminate the critical errors but try and avoid going down a 🐰 hole.

File: src/sqlfluff/rules/L052.py
Comment (barrywhart) [reply src/sqlfluff/rules/L052.py:252]: Really? I think it's helpful -- it explains in the line below why we are switching `lintfix_fn` from `create_after` to `replace`, which otherwise may not be obvious to a future maintainer. Do you think there's a better way to explain what's happening here?

File: test/fixtures/rules/std_rule_cases/L053.yml
Comment (barrywhart) [reply test/fixtures/rules/std_rule_cases/L053.yml:81]: It's because of a YAML limitation. The fix string has a blank line at the end, but the YAML parser doesn't pick it up; it assumes blank line means end of string. The `fail_str` doesn't have a blank line. Think I should change it? I prefer to use "normal" multi-line strings when possible, using quoted strings with `\n` or other escape sequences only when necessary or for readability. Happy to change the `fail_str` if you like, though!

Comment (barrywhart) [reply test/fixtures/rules/std_rule_cases/L053.yml:81]: I'll go ahead and change `fail_str`.

File: src/sqlfluff/rules/L052.py
Comment (tunetheweb) [reply src/sqlfluff/rules/L052.py:252]: Ok see that now. Thought it was an explanation of why you rewrote it the way you did. Resolving

File: src/sqlfluff/rules/L039.py
Comment (tunetheweb) [reply src/sqlfluff/rules/L039.py:132]: Thanks makes sense. Maybe add a comment saying, something like:

This rule works in two steps to remove unnecessary white space:
1. Replace duplicate whitespace to one single whitespace
2. Remove single white spaces if needed.

This can result in two delete being applied to same segment so area so check for that and replace with single delete.

File: src/sqlfluff/core/parser/segments/base.py
Comment (tunetheweb) [reply src/sqlfluff/core/parser/segments/base.py:90]: SGTM

File: test/fixtures/rules/std_rule_cases/L053.yml
Comment (barrywhart) [reply test/fixtures/rules/std_rule_cases/L053.yml:74]: Done!

File: test/fixtures/rules/std_rule_cases/L053.yml
Comment (barrywhart) [reply test/fixtures/rules/std_rule_cases/L053.yml:80]: Done!

**Comment by barrywhart** [comment]
@tunetheweb: Ready for another review.

**Comment by barrywhart** [comment]
@alanmcruickshank: You may be interested in the changes to `BaseSegment.apply_fixes()`. We now handle one case of "overlapping fixes": two fixes with the same anchor, one is `create_before` and one is `create_after`.

@OTooleMichael: You may be interested because it adds a new "sanity check" to the fixes. Previously, rules could return conflicting fixes -- it would apply one fix and silently ignore the rest.

**Pull request merged**
